sql: add the getdatabaseencoding() builtin function

abhishek20123g opened the pull request:
Resolves #41771.

This commit adds builtin function, getdatabaseencoding(),
and the unit-test case for it.

Release note (sql change): This PR is introduced to add builtin
function, getdatabaseencoding(), which returns the current encoding
name used by the database.

abhishek20123g committed:
sql: add the getdatabaseencoding() builtin function

Resolves https://github.com/cockroachdb/cockroach/issues/41771.

This commit adds builtin function, getdatabaseencoding(),
and the unit-test case for it.

Release note (sql change): This PR is introduced to add builtin
function, getdatabaseencoding(), which returns the current encoding
name used by the database.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45129)
<!-- Reviewable:end -->

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=45129) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=45129) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cockroachdb/cockroach?pullRequest=45129) it.</sub>

rohany reviewed (dismissed):
Thanks for the contribution! This needs a test before it can land. You should add a logic test for it (perhaps in `pkg/sql/logictest/testdata/logic_test/pg_builtins`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45129#-:-M09Vdp0DzzZBs3tUxYy:b-dfehww)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @abhishek20123g and @otan)

___
*[pkg/sql/sem/builtins/builtins.go, line 3073 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45129#-M09Ukmq0gLsFnFenGLE:-M09Ukmr19dpYdzCdluQ:b9o3i1o) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e8c6e0c8746f18a72d4971d8dea9caa505e097/pkg/sql/sem/builtins/builtins.go#L3073)):*
> ```Go
> 			ReturnType: tree.FixedReturnType(types.String),
> 			Fn: func(ctx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
> 				rows1, err1 := ctx.InternalExecutor.QueryRow(
> ```

nit: i'd name these just `rows, err`.
___
*[pkg/sql/sem/builtins/builtins.go, line 3081 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45129#-M09V2t_0bZRmqq2Lf4t:-M09V2t_0bZRmqq2Lf4u:bt9ioo4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e8c6e0c8746f18a72d4971d8dea9caa505e097/pkg/sql/sem/builtins/builtins.go#L3081)):*
> ```Go
> 				encodingID, ok := rows1[0].(*tree.DInt)
> 				if !ok {
> 					return tree.DNull, errors.New("failed to type assert Datum to DInt")
> ```

this should return nil rather than `DNull`. Additionally, I'd make this an `errors.AssertionFailure` with more detail. In this case we already know that we couldn't perform the type assertion, but we should have more context on what exactly we expected vs what we found.
___
*[pkg/sql/sem/builtins/builtins.go, line 3083 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45129#-M09VLlX79DFoAD8w-XH:-M09VLlX79DFoAD8w-XI:b2vew7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e8c6e0c8746f18a72d4971d8dea9caa505e097/pkg/sql/sem/builtins/builtins.go#L3083)):*
> ```Go
> 					return tree.DNull, errors.New("failed to type assert Datum to DInt")
> 				}
> 				if err1 != nil {
> ```

You need to check if the error is not nil before accessing the result of the query.
___
*[pkg/sql/sem/builtins/builtins.go, line 3084 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45129#-M09VSd9BzjZ5UQRMtNc:-M09VSd9BzjZ5UQRMtNd:b-axxltc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e8c6e0c8746f18a72d4971d8dea9caa505e097/pkg/sql/sem/builtins/builtins.go#L3084)):*
> ```Go
> 				}
> 				if err1 != nil {
> 					return tree.DNull, err1
> ```

same, this should be nil.
___
*[pkg/sql/sem/builtins/builtins.go, line 3088 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45129#-M09VZKkEQxiQHyja3NG:-M09VZKkEQxiQHyja3NH:bm5ixqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2e8c6e0c8746f18a72d4971d8dea9caa505e097/pkg/sql/sem/builtins/builtins.go#L3088)):*
> ```Go
> 
> 				// Using pg_encoding_to_char() to convert an encoding ID to an encoding name.
> 				rows2, err2 := ctx.InternalExecutor.QueryRow(
> ```

Why not just call `pg_encoding_to_char` in your first query?


<!-- Sent from Reviewable.io -->

abhishek20123g commented:
Hi @rohany 
Thanks for suggesting changes, changes had been made as requested.
I have combined two separate queries for evaluating encoding_id and converting encoding_id to encoding name into one query (which makes more sense).
I am returning nil instead of DNull.
Refactored row1, err1 to row, err, respectively.

wrt Adding test-cases for getdatabaseencoding(),
I had added test-case in '''pkg/sql/logictest/testdata/logic_test/builtin_function''', I am not sure about whether i should add unit test-case in this file or in other file. I have decided to add a test-case in this file because getdatabasencoding() is a builtin function, and most of the functions related to it are also present in this file. And I was only able to add test-case corresponding to UTF8 because perhaps CockroachDB currently only supports UTF8 encoding.

rohany commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -3064,6 +3064,29 @@ may increase either contention or retry errors, or both.`,
+				row, err := ctx.InternalExecutor.QueryRow(
+					ctx.Ctx(), "getdatabaseencoding",
+					ctx.Txn,
+					"SELECT pg_encoding_to_char(encoding) "+
```
Can this not just be `SELECT pg_encoding_to_char(encoding) from pg_database...`? Why do you need a subquery?

abhishek20123g replied:
```select pg_encoding_to_char(encoding) from pg_database where datname = current_database()```
sorry my bad we can work with above commmnd, without creating subquery.

abhishek20123g replied:
```select pg_encoding_to_char(encoding) from pg_database where datname = $1```
sorry my bad we can work with above commmnd, without creating subquery.

I am making these changes too.

rohany commented:
The test looks good! We do only support UTF8 right now.

Some more nits:
* Can you squash these commits into a single commit?
* Can you rename the commit to "sql: add the getdatabaseencoding() builtin function"
* Please use the commit template within cockroach to write a commit message. Importantly, your commit needs a release note. The instructions for this are in the commit template.

abhishek20123g commented:
Absolutely, I will squash all the commit into one, and make them look as recommended.

rohany commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -3064,6 +3064,28 @@ may increase either contention or retry errors, or both.`,
+			Types:      tree.ArgTypes{},
+			ReturnType: tree.FixedReturnType(types.String),
+			Fn: func(ctx *tree.EvalContext, args tree.Datums) (tree.Datum, error) {
+				// computing encoding name bypassing encoding_id which is
```
Can you change this comment to

"Compute encoding name by calling pg_encoding_to_char on the encoding from pg_database."

All comments should be complete sentences!

rohany commented:
Aside from my nit about the comment, this looks good! Once you fix that I will merge the PR.

abhishek20123g commented:
Hi @rohany,
I have changed comment as recommended.

And thank you for guiding me, at various points in this PR. 
For my future PR's, I will try to make them easy to merge.

rohany commented:
Can you rebase this onto master? It looks like you're running into some sort of test flake.

otan commented on pkg/sql/sem/builtins/builtins.go:
```diff
@@ -3064,6 +3064,28 @@ may increase either contention or retry errors, or both.`,
+				row, err := ctx.InternalExecutor.QueryRow(
+					ctx.Ctx(), "getdatabaseencoding",
+					ctx.Txn,
+					"SELECT pg_encoding_to_char(encoding) "+
```
i think this should go in `pg_builtins.go`, and we should probably try re-using 

https://github.com/cockroachdb/cockroach/blob/bb14400cdd1397f47c3781b4830199f169cef497/pkg/sql/sem/builtins/pg_builtins.go#L567

rather than trying to call a nested database query.

abhishek20123g replied:
Sure, I will move it to ```pg_builtins.go```
and you mean rather calling ```pg_encoding_to_char``` inside the query, I should use the same logic in ```pg_encoding_to_char``` to convert encodingID to encoding name.

can we do it by creating some kind of external helper function which converts encodingID to the encoding name which is being called by both, ```pg_encoding_to_char``` and ```getdatabaseencoding```? 
because otherwise, we would have update logics inside both functions for adding support to another encoding type.

otan replied:
> can we do it by creating some kind of external helper function which converts encodingID to the encoding name which is being called by both, pg_encoding_to_char and getdatabaseencoding?

yes that would be my suggestion :)

abhishek20123g replied:
sure, I will update.

otan commented on pkg/sql/logictest/testdata/logic_test/builtin_function:
```diff
@@ -2556,3 +2556,8 @@ SELECT timezone('1970-01-01 01:00'::timestamptz, 'UTC+6')
 SET TIME ZONE +0
+
+query T
+SELECT getdatabaseencoding()
```
can you give this an example of moving to a different encoding and checking it works as expected?

abhishek20123g replied:
I have tried to create, test case for different encoding but not sure how I could create database with different encoding.

otan replied:
according to the docs, you can use `CREATE DATABASE ... ENCODING = x` (see https://www.cockroachlabs.com/docs/stable/create-database.html)

so create that database, use that database, check encoding is same, and ideally reset state back to the original database used (i think it's `test`).

can you also add `subtest getdatabaseencoding` above `query T` to add some sort of separation? thanks!

abhishek20123g replied:
sure, and thanks

otan replied:
looking at the docs again, we only support utf-8 encoding anyway.

abhishek20123g replied:
yes, I am unable to create databases having encoding other than 'UTF-8'

abhishek20123g commented:
hi @otan,
I have moved ```getdatabaseencoding()``` form ```bulitins.go``` to ```pg_builtins.go``` .
Created a helper function called ```getEncodingNameForEncodingID```.
Updated the defination of ```pg_encoding_to_char``` and ```getdatabaseencoding ```.

Do I have to rebase changes to master as suggested by rohany ?

otan commented on pkg/sql/sem/builtins/pg_builtins.go:
```diff
@@ -563,13 +563,33 @@ var pgBuiltins = map[string]builtinDefinition{
+				row, err := ctx.InternalExecutor.QueryRow(
+					ctx.Ctx(), "getdatabaseencoding",
+					ctx.Txn,
+					"SELECT encoding FROM pg_database WHERE datname = $1", ctx.SessionData.Database)
```
sorry bit slow to realise this - looks like we're always returning UTF-8 for this anyway. We populate `pg_database` with a virtual schema, and we always return the same col for this rowtype:
 
https://github.com/cockroachdb/cockroach/blob/bb14400cdd1397f47c3781b4830199f169cef497/pkg/sql/pg_catalog.go#L1084

Perhaps we should also just return UTF-8 here as well. Make a comment on the virtual schema that if we ever change the value of that, we should change it here and vice versa.

Sorry for the round-aboutness with getting this out there.

otan replied:
(note: i would generally try avoid calling another query inside here unless you had to, which led me down this path to finding this :] ideally, if it's configurable, we'd have a function that lets us figure this out without shelling out to another query)

abhishek20123g replied:
So, rather than using InternalExecutor to execute the query you are recommending me to 
just return UTF-8 and add a comment to virtual schema to that if we change value there we update it here too.

like use in ```pg_client_encoding``` in ```pg_builtins.go```

ajwerner commented:
Flaked on #44944.

otan reviewed (approved):
looks good to me - i made a minor jiggle with the comments for you.

thanks for contributing and hope to see more!

otan commented:
bors r=otan

abhishek20123g commented:
thanks, @rohany, and @otan for helping me out 

Can you recommend to me some issues where I could start working?
thanks

rohany commented:
Our good first issue page has a ton of issues that you could take a look at!
